icu4c/source/common/ubidiimp.h

```diff
@@ -198,8 +198,8 @@ typedef struct Run {
 /* in a Run, logicalStart will get this bit set if the run level is odd */
 #define INDEX_ODD_BIT (1UL<<31)
 
-#define MAKE_INDEX_ODD_PAIR(index, level) ((index)|((int32_t)(level)<<31))
-#define ADD_ODD_BIT_FROM_LEVEL(x, level)  ((x)|=((int32_t)(level)<<31))
+#define MAKE_INDEX_ODD_PAIR(index, level) ((index)|((int32_t)(level&1)<<31))
```

Comment:
You need to use ((level)&1) with the extra parentheses so that if the "call" site uses an expression, not just a constant or a variable name, you don't get a syntax error in the macro text substitution.

Same on the next line.

Reply:
I'm not sure this will fix the problem. Which is signed integer overflow. Shifting the 1 bit into the sign bit will still be undefined.

Might be better to forget the masking, and cast the level to unsigned.

Reply:
> You need to use ((level)&1) with the extra parentheses so that if the "call" site uses an expression, not just a constant or a variable name, you don't get a syntax error in the macro text substitution.
> 
> Same on the next line.

well do

Reply:
> I'm not sure this will fix the problem. Which is signed integer overflow. Shifting the 1 bit into the sign bit will still be undefined.

I am sure it will fix the problem since I can reproduce the UBSAN in my system and it does fix it.

> 
> Might be better to forget the masking, and cast the level to unsigned.
According to UBSAN bot is
"runtime error: left shift of 2 by 31 places cannot be represented in type 'int32_t' (aka 'int')"

level is 2 (bit 2 on) and the shift shift it 31 bit left, and the bit 2 is in the location of bit 33- which does not exist.

And that is the problem.

Reply:
Looking into it a little more, shifting a 1 into the sign bit is defined behavior for C++, but is undefined for plain C.
Shifting a 1 out altogether is undefined for both languages.
That is for signed shifts. For unsigned, shifting high order bits out is defined behavior, so long as the total shift amount is less than the width of the variable being shifted.